include/zephyr/posix/sys/utsname.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2023 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include "version.h"
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <zephyr/net/hostname.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* sysname */
+#define SYS_NAMELEN sizeof("Zephyr")
+
+/* nodename */
+#ifdef CONFIG_NET_HOSTNAME_ENABLE
+#define NODE_NAMELEN NET_HOSTNAME_MAX_LEN
+#else
+#define NODE_NAMELEN SYS_NAMELEN
+#endif
+
+/* release */
+#define RELEASE_NAMELEN sizeof(KERNEL_VERSION_STRING)
+
+/* version */
+#if defined(__DATE__) && defined(__TIME__)
+#define DATETIME_LEN (sizeof(__DATE__) + sizeof(__TIME__))
+#else
+#define DATETIME_LEN 0
+#endif
+
+#ifdef BUILD_VERSION
+#define VERSION_BUILD STRINGIFY(BUILD_VERSION)
+#else
+#define VERSION_BUILD KERNEL_VERSION_STRING
+#endif
+#define VERSION_BUILD_LEN sizeof(VERSION_BUILD)
+
+/**
+ * In case anyone wondered if this length has been adjusted for '\0', it will
+ * be formatted in utsname.c as:
+ *     "%s %s %s", VERSION_BUILD, __DATE__, __TIME__
+ * the '\0' of VERSION_BUILD_LEN & __DATE__ will be used for the space that
+ * follows, '\0' of __TIME__ will be used for the \0 of the formatted VERSION
+ * string. In the end the size of the array should fit the string snugly.
+ */
+#define VERSION_NAMELEN (VERSION_BUILD_LEN + DATETIME_LEN)
+
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
```

Comment:
While I appreciate that this is all very precise, I kind of think we would be better off keeping it simple for now.

E.g. add a Kconfig integer option `CONFIG_POSIX_UTSNAME_LENGTH`, and use
```
#ifdef CONFIG_POSIX_UTSNAME
#define _UTSNAME_LENGTH CONFIG_POSIX_UTSNAME_LENGTH
#else
#define _UTSNAME_LENGTH sizeof("Zephyr")
#endif
```

https://man7.org/linux/man-pages/man2/uname.2.html

Inside of the `utsname.c` (which may be better to name `uname.c`), we can have several compile-time `BUILD_ASSERT()` statements. E.g.

```cpp
BUILD_ASSERT(sizeof(z_name.sysname) >= sizeof("Zephyr"));
BUILD_ASSERT(sizeof(z_name.release) >= sizeof(KERNEL_VERSION_STRING));
BUILD_ASSERT(sizeof(z_name.version) >= sizeof(UTSNAME_VERSION(VERSION_BUILD));
BUILD_ASSERT(sizeof(z_name.machine) >= sizeof(CONFIG_ARCH));

```

Reply:
@keith-packard if we remove the `CONFIG_POSIX_UNAME_VERSION_LEN`, then we will have to reinstate the complicated part of the computation here